File: api-java-sql/src/main/java/com/nebulahotel/controller/AvaliacoesController.java
package com.nebulahotel.controller;

import com.nebulahotel.model.Avaliacoes;
import com.nebulahotel.repository.AvaliacoesRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/avaliacoes")
public class AvaliacoesController {

    private final AvaliacoesRepository avaliacoesRepository;

    public AvaliacoesController(AvaliacoesRepository avaliacoesRepository) {
        this.avaliacoesRepository = avaliacoesRepository;
    }

    @GetMapping
    public List<Avaliacoes> listarTodas() {
        return avaliacoesRepository.findAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Avaliacoes> buscarPorId(@PathVariable Integer id) {
        return avaliacoesRepository.findById(id)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @GetMapping("/resumo")
    public Map<String, Object> resumoAvaliacoes() {
        return Map.of(
                "totalAvaliacoes", avaliacoesRepository.count(),
                "mediaGeral", avaliacoesRepository.calcularMediaGeral()
        );
    }

    @PostMapping
    public ResponseEntity<?> criar(@RequestBody Avaliacoes avaliacao) {
        if (avaliacao.getIdFeedback() == null) {
            return ResponseEntity.badRequest().body(Map.of("erro", "O campo idFeedback é obrigatório."));
        }

        if (avaliacoesRepository.existsById(avaliacao.getIdFeedback())) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(Map.of("erro", "Já existe avaliação com este idFeedback."));
        }

        if (avaliacao.getDataAvaliacao() == null) {
            avaliacao.setDataAvaliacao(LocalDateTime.now());
        }

        Avaliacoes salvo = avaliacoesRepository.save(avaliacao);
        return ResponseEntity.status(HttpStatus.CREATED).body(salvo);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Avaliacoes> atualizar(@PathVariable Integer id, @RequestBody Avaliacoes avaliacaoAtualizada) {
        if (!avaliacoesRepository.existsById(id)) {
            return ResponseEntity.notFound().build();
        }

        avaliacaoAtualizada.setIdFeedback(id);
        Avaliacoes salvo = avaliacoesRepository.save(avaliacaoAtualizada);
        return ResponseEntity.ok(salvo);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> remover(@PathVariable Integer id) {
        if (!avaliacoesRepository.existsById(id)) {
            return ResponseEntity.notFound().build();
        }

        avaliacoesRepository.deleteById(id);
        return ResponseEntity.noContent().build();
    }
}

File: api-kotlin-nosql/src/main/kotlin/com/nebulahotel/controller/AvaliacoesController.kt
package com.nebulahotel.controller

import com.nebulahotel.model.Avaliacoes
import com.nebulahotel.repository.AvaliacoesRepository
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal
import java.math.RoundingMode

@CrossOrigin
@RestController
@RequestMapping("/avaliacoes")
class AvaliacoesController(
    private val avaliacoesRepository: AvaliacoesRepository
) {

    @GetMapping
    fun listarTodas(): List<Avaliacoes> = avaliacoesRepository.findAll()

    @GetMapping("/{id}")
    fun buscarPorId(@PathVariable id: Int): ResponseEntity<Avaliacoes> {
        val avaliacao = avaliacoesRepository.findById(id)
        return if (avaliacao.isPresent) ResponseEntity.ok(avaliacao.get()) else ResponseEntity.notFound().build<Avaliacoes>()
    }

    @GetMapping("/resumo")
    fun resumoAvaliacoes(): Map<String, Any> {
        val avaliacoes = avaliacoesRepository.findAll()
        val total = avaliacoes.size
        val media = if (total == 0) {
            BigDecimal.ZERO
        } else {
            avaliacoes
                .map { it.notaGeral }
                .reduce(BigDecimal::add)
                .divide(BigDecimal(total), 2, RoundingMode.HALF_UP)
        }

        return mapOf(
            "totalAvaliacoes" to total,
            "mediaGeral" to media
        )
    }

    @PostMapping
    fun criar(@RequestBody avaliacao: Avaliacoes): ResponseEntity<Any> {
        if (avaliacoesRepository.existsById(avaliacao.idFeedback)) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                .body(mapOf("erro" to "Já existe avaliação com este idFeedback."))
        }
        return ResponseEntity.status(HttpStatus.CREATED).body(avaliacoesRepository.save(avaliacao))
    }

    @PutMapping("/{id}")
    fun atualizar(@PathVariable id: Int, @RequestBody avaliacaoAtualizada: Avaliacoes): ResponseEntity<Avaliacoes> {
        if (!avaliacoesRepository.existsById(id)) {
            return ResponseEntity.notFound().build<Avaliacoes>()
        }
        val salvo = avaliacoesRepository.save(avaliacaoAtualizada.copy(idFeedback = id))
        return ResponseEntity.ok(salvo)
    }

    @DeleteMapping("/{id}")
    fun remover(@PathVariable id: Int): ResponseEntity<Void> {
        if (!avaliacoesRepository.existsById(id)) {
            return ResponseEntity.notFound().build()
        }
        avaliacoesRepository.deleteById(id)
        return ResponseEntity.noContent().build()
    }
}
